core/model-api/src/main/java/org/eclipse/rdf4j/model/base/CoreDatatype.java

```diff
@@ -0,0 +1,340 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Eclipse RDF4J contributors.
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Distribution License v1.0
+ *  which accompanies this distribution, and is available at
+ *  http://www.eclipse.org/org/documents/edl-v10.php.
+ ******************************************************************************/
+
+package org.eclipse.rdf4j.model.base;
+
+import java.io.Serializable;
+import java.util.Optional;
+
+import javax.xml.datatype.Duration;
+import javax.xml.datatype.XMLGregorianCalendar;
+
+import org.eclipse.rdf4j.model.IRI;
+
+public interface CoreDatatype {
+
+	CoreDatatype NONE = () -> {
+		throw new IllegalStateException();
+	};
+
+	/**
+	 * Checks whether the supplied datatype is an XML Schema Datatype.
+	 *
+	 * @return true if the datatype is an XML Schema Datatype
+	 */
+	default boolean isXSDDatatype() {
+		return false;
+	}
+
+	default boolean isRDFDatatype() {
+		return false;
+	}
+
+	default boolean isGEODatatype() {
+		return false;
+	}
+
+	default XSD asXSDDatatypeOrNull() {
+		return null;
+	}
+
+	default RDF asRDFDatatypeOrNull() {
+		return null;
+	}
+
+	default GEO asGEODatatypeOrNull() {
+		return null;
+	}
```

Comment:
Can we remove the `OrNull` bit from the method names here and return an Optional instead? We could then do:

```java
CoreDatatype.XSD xsdDatatype = yourLiteral.getCoreDatatype().asXSDDatatype().orElse(null);
```

which IMHO is a bit cleaner.

Reply:
I kinda wanted to stay away from using Optional. Performance wasn't great and probably won't be until GraalVM becomes the norm. I would rather return null in this case, but I can change the method name and add some javadocs.

Reply:
Let me double check the performance though. I have a branch where I've rewritten the sorting code to use CoreDatatype, so I can check there if it's actually any slower.

Reply:
Performance was actually perfectly fine.